pkg/handler/collector/deps_dev/deps_dev.go

```diff
@@ -133,6 +135,17 @@ func (d *depsCollector) populatePurls(ctx context.Context, docChannel chan<- *pr
 	if err != nil {
 		return fmt.Errorf("unable to retrieve datasource: %w", err)
 	}
+
+	if !d.retrieveDependencies {
+		for _, purl := range ds.PurlDataSources {
+			err = d.retrieveMetadataWithoutDeps(ctx, purl, docChannel)
```

Comment:
could this be made into a non-blocking go routine? That would speed up the fetching as they should all be independent of each other.

Reply:
Making each query in a separate goroutine, like this

``` 
for _, purl := range ds.PurlDataSources {
	go func(p datasource.Source) {
		d.retrieveMetadataWithoutDeps(ctx, p, docChannel)
	}(purl)
}
```
will cause the call to `populatePurls`, and then `RetrieveArtifacts` (at least in non-poll mode), to return without blocking. I think this breaks the contract of `RetrieveArtifacts` [here](https://github.com/guacsec/guac/blob/e48e534763f7d70750141e2a9d1047b8caea812c/pkg/handler/collector/collector.go#L42), that the function should block until all of the artifacts are collected.

Reply:
create wait groups and block until it is all collected?

Reply:
Added this, thanks!

Reply:
The nested calls mutate state, and it isn't thread-safe.

Reply:
Ah true, thanks @naveensrinivasan. Should I revert this change or add synchronization? Making these calls concurrently was much faster, but making the calls sequentially is similar to how dependencies are fetched sequentially in [getAllDependencies](https://github.com/guacsec/guac/blob/51e8fc6212ac88a8d53d30922520690b8fb97a74/pkg/handler/collector/deps_dev/deps_dev.go#L154), for reference.

Reply:
The previous depsdev PR(https://github.com/guacsec/guac/pull/1263) was essentially to speed up the fetching of blocking calls. 

If we go with sequential calls, things will be slow.

Adding sync is not recommended in most conditions to avoid future deadlocks and race conditions. 

I recommend refactoring with channels, avoiding the global state returning values from these calls, and then mutating the global state.

Reply:
I changed this to first concurrently retrieve all of the project and version information and once finished, generate the metadata documents with that information. The goroutines don't mutate any shared state, and they work in the same way as in `getAllDependencies()`

Reply:
Thanks @naveensrinivasan and @mdeicas! This looks great now!